File: backend/app/schemas/health_card.py
"""
Pydantic v2 schemas for HealthCard request/response serialization.

Security notes
--------------
- No schema in this file includes PHI beyond the minimal PatientVerifySummary
  fields (patient_code, full_name, age, sex, priority flags).
- The QR payload and NFC payload schemas encode ONLY patient_id and
  card_version — never name, DOB, diagnosis, or any other PHI.
- HealthCardResponse includes qr_data_uri only on generate/reissue responses;
  it is NOT included on metadata-only GET responses (frontend regenerates
  QR preview via qrGenerator.worker.ts using patient_id + card_version).

SDP Reference: Section 6.6
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.schemas._base import BaseSchema


# ---------------------------------------------------------------------------
# Request schemas
# ---------------------------------------------------------------------------


class NfcLinkRequest(BaseSchema):
    """
    POST /health-cards/{patient_id}/nfc-link

    Body sent by the BHW when provisioning a physical NFC chip.
    """

    nfc_uid: str = Field(
        ...,
        min_length=1,
        max_length=64,
        description="Hardware UID of the physical NFC chip (e.g. '04:1A:2B:3C:4D:5E:6F')",
    )


class CardVerifyRequest(BaseSchema):
    """
    POST /health-cards/verify

    The frontend sends either qr_payload (scanned URL string) or nfc_uid
    (from a physical tap) — or both.  At least one field must be present.

    Security note: never echo back the qr_payload or nfc_uid in the response.
    """

    qr_payload: str | None = Field(
        None,
        description="Raw QR payload URL string (contains only patient_id, card_version, sig)",
    )
    nfc_uid: str | None = Field(
        None,
        max_length=64,
        description="Hardware UID read from the tapped NFC chip",
    )


# ---------------------------------------------------------------------------
# Response schemas
# ---------------------------------------------------------------------------


class HealthCardResponse(BaseSchema):
    """
    Returned by generate, reissue, get, and nfc-link endpoints.

    qr_data_uri is only populated by generate and reissue responses; it
    contains the base64 QR image so the frontend can display it immediately
    without a second round-trip.  GET /health-cards/{id} omits it (None).

    NEVER add name, address, birth_date, philhealth_no, or diagnosis here.
    """

    id: str
    patient_id: str
    card_number: str
    card_version: int
    status: Literal["active", "lost", "reissued", "revoked"]
    issued_at: datetime
    expires_at: datetime | None = None
    nfc_uid: str | None = None

    # Populated only on generate/reissue (not on GET metadata requests)
    qr_data_uri: str | None = Field(
        None,
        description="data:image/png;base64,... QR code image — only present on card generation/reissue",
    )


class CardGenerateResponse(BaseSchema):
    """
    Full response for POST .../generate and POST .../reissue.

    Bundles card metadata, the QR image data URI (for immediate display),
    and the NFC payload dict (to write to the physical chip).
    """

    card: HealthCardResponse
    # base64 PNG data URI — display immediately, no second round-trip needed
    qr_data_uri: str = Field(
        description="data:image/png;base64,... — QR code for the card front"
    )
    # Minimal JSON to write to the NFC chip
    nfc_payload: dict[str, str | int] = Field(
        description="{'patient_id': str, 'card_version': int} — write to NFC chip NDEF record"
    )


class PatientVerifySummary(BaseSchema):
    """
    Returned by POST /health-cards/verify on successful card verification.

    This is the ONLY patient data the verify endpoint ever returns.
    Deliberately minimal — front-desk staff need to confirm identity and
    see priority flags, nothing more.

    Fields intentionally omitted:
      - address, guardian info, mobile number (not needed at the desk)
      - philhealth_no (sensitive financial data)
      - medical_history, diagnosis (clinical PHI)
      - birth_date (derivable from age — not needed at this screen)
    """

    patient_code: str = Field(description="e.g. BHC-2026-000042")
    full_name: str = Field(description="Formatted: 'First [Middle] Last'")
    age: int = Field(ge=0, description="Age in whole years")
    sex: str = Field(description="'male' or 'female'")
    is_senior: bool = Field(description="Age ≥ 60 — priority queue flag")
    is_pwd: bool = Field(description="Person with Disability — priority queue flag")
    is_pregnant: bool = Field(description="Current pregnancy status")
    last_visit_date: datetime | None = Field(
        None, description="Timestamp of the most recent visit record, or None"
    )
    card_status: str = Field(description="'active', 'lost', 'reissued', or 'revoked'")

File: frontend/web/types/healthCard.ts
/**
 * Health card TypeScript interfaces.
 * Mirrors backend/app/schemas/health_card.py
 *
 * Security note: No PHI beyond patient_code, full_name, age, sex, and
 * priority flags appears in any card-related type.  HealthCardData is a
 * pointer — it identifies the card, not the patient's medical information.
 */

// ---------------------------------------------------------------------------
// Card status
// ---------------------------------------------------------------------------

export type CardStatus = "active" | "lost" | "reissued" | "revoked";

// ---------------------------------------------------------------------------
// Core health card record (mirrors HealthCardResponse schema)
// ---------------------------------------------------------------------------

export interface HealthCardData {
  id: string;
  patient_id: string;
  card_number: string;
  card_version: number;
  status: CardStatus;
  issued_at: string; // ISO datetime string
  expires_at: string | null;
  nfc_uid: string | null;
  /**
   * base64 PNG data URI — only present on generate/reissue responses.
   * null on GET metadata responses (frontend regenerates from patient_id + version).
   */
  qr_data_uri?: string | null;
}

// ---------------------------------------------------------------------------
// Generate / reissue response (mirrors CardGenerateResponse schema)
// ---------------------------------------------------------------------------

export interface CardGenerateResponse {
  card: HealthCardData;
  /** data:image/png;base64,... — QR image for immediate display. */
  qr_data_uri: string;
  /** Minimal JSON to write to the NFC chip via lib/nfc.ts writeNfcTag(). */
  nfc_payload: NfcPayload;
}

// ---------------------------------------------------------------------------
// NFC chip payload (mirrors nfc_payload_service.build_nfc_payload())
// ---------------------------------------------------------------------------

/**
 * The ONLY data written to the physical NFC chip.
 * Never includes PHI — the chip is a pointer to a patient record.
 */
export interface NfcPayload {
  patient_id: string;
  card_version: number;
}

// ---------------------------------------------------------------------------
// Verify endpoint request
// ---------------------------------------------------------------------------

export interface CardVerifyRequest {
  /** Raw QR payload URL string (contains only patient_id, card_version, sig). */
  qr_payload?: string;
  /** Hardware UID read from the tapped NFC chip. */
  nfc_uid?: string;
}

// ---------------------------------------------------------------------------
// Verify endpoint response (mirrors PatientVerifySummary schema)
// ---------------------------------------------------------------------------

/**
 * Minimal patient summary returned by POST /health-cards/verify.
 *
 * Deliberately excludes: address, mobile_number, philhealth_no, birth_date,
 * guardian info, diagnosis, treatment_notes, and all other PHI.
 *
 * The front-desk screen needs ONLY these fields to confirm identity and
 * route to the priority queue.
 */
export interface PatientVerifySummary {
  patient_code: string;
  full_name: string;
  age: number;
  sex: string;
  is_senior: boolean;
  is_pwd: boolean;
  is_pregnant: boolean;
  last_visit_date: string | null;
  card_status: CardStatus;
}

// ---------------------------------------------------------------------------
// Extended verify response — returned when ?full=true (authenticated staff)
// ---------------------------------------------------------------------------

/**
 * Extended patient summary returned by POST /health-cards/verify?full=true.
 *
 * Adds patient_id, birth_date, mobile_number, and photo_url to the base
 * PatientVerifySummary.  A PHI_VIEW audit log entry is written server-side
 * when this response is issued.
 *
 * photo_url is a relative path ("/media/patient_photos/<uuid>.jpg") that
 * must be prefixed with the API base host to form an absolute URL.
 */
export interface PatientVerifySummaryFull extends PatientVerifySummary {
  /** UUID string — used to build navigation links to the full patient record. */
  patient_id: string;
  /** ISO date "YYYY-MM-DD" */
  birth_date: string;
  mobile_number: string | null;
  /**
   * Relative path "/media/patient_photos/<uuid>.jpg" or null.
   * Prefix with API host (e.g. http://192.168.100.6:8000) to form absolute URL.
   */
  photo_url: string | null;
}

// ---------------------------------------------------------------------------
// NFC link request
// ---------------------------------------------------------------------------

export interface NfcLinkRequest {
  nfc_uid: string;
}
